### PR TITLE
feat: make IAM credentials expiry window customizable

### LINKED ISSUE #2136: Make AWS IAM Credentials Expiry Window Customizable
Hi 👋

This is something of a duplicate / reopen of #1075

I'd also really like to have the expiry window for the AWS IAM window be customizable.

We (at https://github.com/secureCodeBox/secureCodeBox) use the minio client to create presigned url for objects in the s3 buckets.
People using the AWS IAM authentication method often run into problems when the token lifetime is coming to the end and a presigned url is created that has a longer lifetime than the token. AWS then seems to cap the lifecycle of the presigned url to the lifetime of the token (which makes sense)

So for us it would be ideal if we had some way to configure it from the default of 15min, say 1h or 4h so that the presigned urls have a longer min. lifetime than the 15min.

If thats cool with you'all I'm happy to try to open up a PR for it.

### PR DESCRIPTION
**Pull request by allanrogerr**
## Description

Adds an exported `ExpiryWindow time.Duration` field to the `IAM` credentials provider so callers can control how long before actual expiration credentials are refreshed.

- Zero value keeps the existing default (`DefaultExpiryWindow`, the 80%-of-lifetime rule) — fully backwards compatible.
- Positive value refreshes that long before actual expiration.
- A window that meets or exceeds the credential lifetime forces a refresh on every retrieval.
- Negative value behaves like the default.
- Applied on both credential paths: web-identity (STS) and IMDS/ECS.
- The web-identity path applies the window to the **raw** STS credential expiration. Previously that path fed the inner provider's already-`DefaultExpiryWindow`-reduced expiration back into `SetExpiration`, compounding the default window (refresh at ~64% of remaining lifetime instead of the documented 80%) and skewing custom windows; it now mirrors the IMDS path.

Fixes #2136

## Motivation and Context

The refresh window was hardcoded to `DefaultExpiryWindow`, with no way to customize it. Users generating presigned URLs under AWS IAM auth hit problems near token end-of-life: AWS caps a presigned URL's lifetime to the remaining token lifetime, so URLs silently expire earlier than requested. A configurable window lets callers refresh credentials earlier and guarantee a longer minimum presigned-URL lifetime.

## How to test this PR?

`go test -race -run ExpiryWindow ./pkg/credentials/`

New tests cover: a custom window (expiry math + boundary), the zero-value default path, a window exceeding the credential lifetime, an explicitly-set default window, an arbitrary negative window (equivalent to the default), and a custom window through the web-identity (STS) path — the last one fails against the pre-fix double-application and passes with the raw-expiration fix.

### Side-by-side proof (base vs. fix)

**Before (pre-fix `iam_aws.go`, new tests present):** the capability does not exist — build fails.

```
# github.com/minio/minio-go/v7/pkg/credentials [github.com/minio/minio-go/v7/pkg/credentials.test]
pkg/credentials/iam_aws_test.go:429:3: unknown field ExpiryWindow in struct literal of type IAM
pkg/credentials/iam_aws_test.go:487:7: p.ExpiryWindow undefined (type *IAM has no field or method ExpiryWindow)
pkg/credentials/iam_aws_test.go:488:73: p.ExpiryWindow undefined (type *IAM has no field or method ExpiryWindow)
pkg/credentials/iam_aws_test.go:498:3: unknown field ExpiryWindow in struct literal of type IAM
pkg/credentials/iam_aws_test.go:521:3: unknown field ExpiryWindow in struct literal of type IAM
FAIL	github.com/minio/minio-go/v7/pkg/credentials [build failed]
```

**After (this PR):** all cases pass.

```
=== RUN   TestIAMCustomExpiryWindow
--- PASS: TestIAMCustomExpiryWindow (0.02s)
=== RUN   TestIAMZeroExpiryWindowKeepsDefault
--- PASS: TestIAMZeroExpiryWindowKeepsDefault (0.01s)
=== RUN   TestIAMExpiryWindowExceedsLifetime
--- PASS: TestIAMExpiryWindowExceedsLifetime (0.01s)
=== RUN   TestIAMExplicitDefaultExpiryWindow
--- PASS: TestIAMExplicitDefaultExpiryWindow (0.01s)
=== RUN   TestIAMNegativeExpiryWindowBehavesLikeDefault
--- PASS: TestIAMNegativeExpiryWindowBehavesLikeDefault (0.02s)
=== RUN   TestIAMCustomExpiryWindowWebIdentity
--- PASS: TestIAMCustomExpiryWindowWebIdentity (0.01s)
PASS
ok  	github.com/minio/minio-go/v7/pkg/credentials	1.145s
```

Full package (`go test -race ./pkg/credentials/`) and `go vet ./pkg/credentials/` both clean — no regressions.

## Test summary: run against BOTH AIStor master and community MinIO master

The `ExpiryWindow` behavior itself is client-side (IAM IMDS/ECS and web-identity credential refresh) and is definitively covered by the unit tests above, which drive a mock IMDS/STS endpoint. To confirm the SDK carrying this change has no regression against either server edition, the identical test matrix was run against a live **AIStor master (licensed)** server and a live **community MinIO master** server — not community only.

**Servers under test**

| Edition | Source | Build version | License |
|---|---|---|---|
| AIStor master | `miniohq/aistor` @ `aea948d17` | `DEVELOPMENT.2026-07-03T05-41-35Z` | MinIO Enterprise License (started with `MINIO_LICENSE` from the mint `minio.license`) |
| Community master | `minio/minio` @ `7aac2a2` | `DEVELOPMENT.2026-02-12T20-18-48Z` | GNU AGPLv3 |

Both single-node, `ENABLE_HTTPS=0`, `MINT_MODE=full`, `SERVER_ENDPOINT=localhost:9000`.

**Results (identical commands on each server)**

| Test set | Command | AIStor master | Community master |
|---|---|---|---|
| IAM credential unit tests (incl. 4 new `ExpiryWindow` cases), race | `go test -race -run 'ExpiryWindow\|TestIAM' ./pkg/credentials/` | ✅ 10 PASS / 0 FAIL | ✅ 10 PASS / 0 FAIL |
| SDK core live tests (`core_test.go`) | `SERVER_ENDPOINT=… go test -count=1 .` | ✅ 78 PASS / 0 FAIL | ✅ 78 PASS / 0 FAIL |
| Functional (mint) suite | `SERVER_ENDPOINT=… MINT_MODE=full ./functional_tests` | ✅ 143 PASS / 0 FAIL / 28 NA | ✅ 109 PASS / 0 FAIL / 30 NA |

`NA` = capability not applicable to a single-node, no-KMS deployment (e.g. SSE-S3 copy-part). Note: this matrix was captured before the two follow-up unit tests (negative window, web-identity custom window) were added; those are covered by the unit-test evidence above at the current head. AIStor exercises more test cases (143 vs 109 PASS) reflecting its broader feature set; **zero failures on either edition**.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Cleanup/Maintenance (no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [x] Internal documentation updated (godoc on the new field)
- [ ] Create a documentation update request [here](https://github.com/miniohq/aistor-object-store-docs/issues/new)
- [x] No internode changes / version interoperability introduced

## Summary by CodeRabbit

* **New Features**
  * Added configurable credential expiration windows for IAM credentials.
  * Supports custom refresh windows, with default behavior retained for zero or negative values.
  * Applies expiration handling consistently across standard and web identity credentials.
  * Credentials with windows exceeding their lifetime are refreshed immediately.

* **Tests**
  * Added coverage for custom, default, negative, oversized, and web identity expiration windows.

### CODE REVIEW

File: pkg/credentials/iam_aws_test.go [hunk 1]
```diff
@@ -419,3 +419,189 @@ func TestIMDSv1Blocked(t *testing.T) {
...
+	}
+	defer os.Remove(f.Name())
+	f.Write([]byte("token"))
+	f.Close()
+
+	t.Setenv("AWS_WEB_IDENTITY_TOKEN_FILE", f.Name())
+	t.Setenv("AWS_ROLE_ARN", "arn:aws:sts::123456789012:assumed-role/FederatedWebIdentityRole/app1")
+	creds, err := p.RetrieveWithCredContext(defaultCredContext)
+	os.Unsetenv("AWS_WEB_IDENTITY_TOKEN_FILE")
+	os.Unsetenv("AWS_ROLE_ARN")
```
[the remark this answers is not in the document]
Comment (allanrogerr) [reply pkg/credentials/iam_aws_test.go]: Fixed in be1b80a — Write/Close errors are now checked with t.Fatal, and the redundant os.Remove/os.Unsetenv cleanup is dropped since t.TempDir and t.Setenv handle both automatically.

File: pkg/credentials/iam_aws.go [hunk 2]
```diff
@@ -94,6 +105,15 @@ func NewIAM(endpoint string) *Credentials {
...
 
+// expiryWindow returns the configured ExpiryWindow, falling back to
+// DefaultExpiryWindow when unset.
+func (m *IAM) expiryWindow() time.Duration {
+	if m.ExpiryWindow == 0 {
+		return DefaultExpiryWindow
+	}
+	return m.ExpiryWindow
+}
+
```
[the remark this answers is not in the document]
Comment (allanrogerr) [reply pkg/credentials/iam_aws.go:116]: Fixed in 73bc7f3 — expiryWindow() now maps all non-positive values to DefaultExpiryWindow, so the documented "negative behaves like the default" contract is guaranteed locally rather than depending on SetExpiration's handling of negative windows (behavior is unchanged today since DefaultExpiryWindow is itself the negative sentinel).

**Review by taran-p — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]
